hudi-common/src/main/java/org/apache/hudi/common/table/read/HoodieBaseFileGroupRecordBuffer.java

```diff
@@ -322,11 +322,13 @@ protected Option<DeleteRecord> doProcessNextDeletedRecord(DeleteRecord deleteRec
             return Option.empty();
           }
           Comparable deleteOrderingVal = deleteRecord.getOrderingValue();
-          // Checks the ordering value does not equal to 0
-          // because we use 0 as the default value which means natural order
-          boolean chooseExisting = !deleteOrderingVal.equals(0)
-              && ReflectionUtils.isSameClass(existingOrderingVal, deleteOrderingVal)
-              && existingOrderingVal.compareTo(deleteOrderingVal) > 0;
+
+          // Because the delete record from older log block, so we should deal it with three cases:
+          // 1. delete record with invalid ordering value: keep existing record because it is newer, deletion can be ignored.
+          // 2. delete record with valid ordering value and higher than existing record: keep delete record because it is newer.
+          // 3. delete record with valid ordering value and lower than existing record: keep existing record because it is newer, deletion can be ignored.
+          boolean chooseExisting = deleteOrderingVal.equals(0)
```

Comment:
this is still not correct, because the delete block lost the processing time sequence semantics, cc @yihua , how do we handle the deletes with processing time sequence semantics now? It looks like we can not cover this scenario because the delete block lost the info of processing time sequence(a delete paylod always has higher priority than the data block payload in this processing time merging scenario).

Reply:
> this is still not correct, because the delete block lost the processing time sequence semantics, cc @yihua , how do we handle the deletes with processing time sequence semantics now? It looks like we can not cover this scenario because the delete block lost the info of processing time sequence(a delete paylod always has higher priority than the data block payload in this processing time merging scenario).

Agree. I think we need to display the instant time in hoodie record which will provide the basis for us to handle delete record and more complex scenarios in the future

Reply:
In addition, there is another question that makes me confused. The concept of natural order is hidden instead of provided. At present, it is assumed that the read order of log block is natural order or `commit_time`.  Is it necessary to put `commit_time` directly into HoodieRecord? It looks like we can get this field out of the data block header.

Reply:
I discussed this issue with @linliu-code and he has a fix.

Reply:
@linliu-code could you link you PR?

Reply:
@TheR1sing3un Maybe you can investigate if we can compare the records with metadata field `HoodieMetadataField.COMMIT_SEQNO_METADATA_FIELD` when they have the same `orderingValue`.

Reply:
> @TheR1sing3un Maybe you can investigate if we can compare the records with metadata field `HoodieMetadataField.COMMIT_SEQNO_METADATA_FIELD` when they have the same `orderingValue`.

I think this is feasible, as long as the transaction time of the data can be known during the merge process, that is, instantTime, then the uncertainty in the processing of delete record can be eliminated